yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -336,94 +349,160 @@ export class AvmEphemeralForest {
+   * If the key is not found in the tree, it does an external lookup to the underlying merkle DB.
+   * Indicates whethe the sibling path is absent in the ephemeral tree.
+   * @param treeId - The tree we are looking up in
+   * @param key - The key for which we are look up the leaf or low leaf for.
+   * @param T - The type of the preimage (PublicData or Nullifier)
+   * @returns [
+   *     preimageWitness - The leaf or low leaf info (preimage & leaf index),
+   *     pathAbsentInEphemeralTree - whether its sibling path is absent in the ephemeral tree (useful during insertions)
+   * ]
+   */
+  async _getLeafOrLowLeafInfo<ID extends IndexedTreeId, T extends IndexedTreeLeafPreimage>(
+    treeId: ID,
+    key: Fr,
+  ): Promise<[PreimageWitness<T>, /*pathAbsentInEphemeralTree=*/ boolean]> {
+    const bigIntKey = key.toBigInt();
+    // In this function, "min" refers to the leaf with the
+    // largest key <= the specified key in the indexedUpdates.
+    // In other words, the leaf with the "next lowest" key in indexedUpdates.
+
+    // First, search the indexed updates (no DB fallback) to find
+    // the leafIndex of the leaf with the largest key <= the specified key.
+    const minIndexedLeafIndex = this._getLeafIndexOrNextLowestInIndexedUpdates(treeId, key);
+    if (minIndexedLeafIndex === -1n) {
+      // No leaf is present in the indexed updates that is <= the key,
+      // so retrieve the leaf or low leaf from the underlying DB.
+      const leafOrLowLeafPreimage: PreimageWitness<T> = await this._getLeafOrLowLeafWitnessInExternalDb(
+        treeId,
+        bigIntKey,
+      );
+      return [leafOrLowLeafPreimage, /*pathAbsentInEphemeralTree=*/ true];
+    } else {
+      // A leaf was found in the indexed updates that is <= the key
+      const minPreimage: T = this.getIndexedUpdate(treeId, minIndexedLeafIndex);
+      if (minPreimage.getKey() === bigIntKey) {
+        // the index found is an exact match, no need to search further
+        const leafInfo = { preimage: minPreimage, index: minIndexedLeafIndex, update: true };
+        return [leafInfo, /*pathAbsentInEphemeralTree=*/ false];
+      } else {
+        // We are starting with the leaf with largest key <= the specified key
+        // Starting at that "min leaf", search for specified key in both the indexed updates
+        // and the underlying DB. If not found, return its low leaf.
+        const leafOrLowLeafInfo = await this._searchForLeafOrLowLeaf<ID, T>(
+          treeId,
+          bigIntKey,
+          minPreimage,
+          minIndexedLeafIndex,
+        );
+        // We did not find it - this is unexpected... the leaf OR low leaf should always be present
+        assert(leafOrLowLeafInfo !== undefined, 'Could not find leaf or low leaf. This should not happen!');
+        return [leafOrLowLeafInfo, /*pathAbsentInEphemeralTree=*/ false];
```

Comment:
I dont think it's safe to assume false here as `this._searchForLeafOrLowLeaf` has the potential to look into the external DB when walking the linked list forward

Reply:
Hmm.... The point of this boolean is "would a write/append need to insert the path?" It is true only in the case when `(minIndexedLeafIndex === -1n)` because that is the only case that triggered an insertion of the absent path in the original function.

Maybe it's just named poorly? I don't think I changed any logic, just created this bool and use it to move the insertion of the missing path into the write/append function.

Reply:
`minIndexedLeafIndex` just means the leaf is smaller than any leaf we have seen (which means that it is definitely not in the set)

But when `minIndexedLeafIndex !== -1n`,  you might find a path that the ephemeral tree doesnt have when calling `this._searchForLeafOrLowLeaf ` .

For example: 
1) The ephemeral tree stores a single leaf:  `Leaf A`,
2) We then look for the low leaf of a new leaf: `Leaf B`,  and we assume `B.getKey() > A.getKey()`
3) It makes sense to start the search from `Leaf A` (since it's more optimal than starting from index 0, so `minIndexedLeafIndex === 0n`)
4) However, in `this._searchForLeafOrLowLeaf ` we will ask for `Leaf A.nextIndex()`, that next index might require looking in the external DB.
5) If we find a `Leaf C = LeafA.nextIndex()` in the external DB AND it is the low leaf we are looking for, we need to store the sibling path of `Leaf C` because we wont have seen it before.

Reply:
Makes sense. I didn't think I had changed base functionality, only moved things into helper functions.

I fixed this in your PR, got tests passing, and merged. Should be good now!